File: backend/app/utils/logger.py
"""
Enhanced logging utilities with structured error reporting
"""
import os
import sys
import traceback
import logging
from datetime import datetime
from functools import wraps
from flask import current_app, request, g
import uuid

class StructuredLogger:
    @staticmethod
    def log_error(logger, error, context=None, user_id=None, request_id=None):
        """Log error with structured information"""
        error_id = str(uuid.uuid4())[:8]
        
        error_data = {
            'error_id': error_id,
            'timestamp': datetime.utcnow().isoformat(),
            'error_type': type(error).__name__,
            'error_message': str(error),
            'user_id': user_id or getattr(request, 'current_user_id', None),
            'request_id': request_id or getattr(g, 'request_id', None),
            'context': context or {},
        }
        
        # Add request info if available
        if request:
            error_data.update({
                'method': request.method,
                'url': request.url,
                'remote_addr': request.remote_addr,
                'user_agent': request.headers.get('User-Agent', 'Unknown')
            })
        
        # Add full traceback in development
        if os.environ.get('FLASK_ENV', 'development') == 'development':
            error_data['traceback'] = traceback.format_exc()
        
        logger.error(f"Error {error_id}: {error_data}")
        return error_id
    
    @staticmethod 
    def log_warning(logger, message, context=None, user_id=None):
        """Log warning with context"""
        warning_data = {
            'timestamp': datetime.utcnow().isoformat(),
            'message': message,
            'user_id': user_id or getattr(request, 'current_user_id', None),
            'context': context or {},
        }
        
        if request:
            warning_data.update({
                'method': request.method,
                'url': request.url,
                'remote_addr': request.remote_addr
            })
            
        logger.warning(f"Warning: {warning_data}")
    
    @staticmethod
    def log_info(logger, message, context=None, user_id=None):
        """Log info with context"""
        info_data = {
            'timestamp': datetime.utcnow().isoformat(), 
            'message': message,
            'user_id': user_id or getattr(request, 'current_user_id', None),
            'context': context or {},
        }
        
        logger.info(f"Info: {info_data}")

def log_route_errors(f):
    """Decorator to automatically log route errors with context"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        try:
            return f(*args, **kwargs)
        except Exception as e:
            error_id = StructuredLogger.log_error(
                current_app.logger,
                e,
                context={
                    'route': f.__name__,
                    'args': args,
                    'kwargs': kwargs
                }
            )
            
            # Return structured error response
            if os.environ.get('FLASK_ENV', 'development') == 'development':
                return {
                    'error': 'Internal server error',
                    'error_id': error_id,
                    'debug': str(e),
                    'type': type(e).__name__
                }, 500
            else:
                return {
                    'error': 'Internal server error', 
                    'error_id': error_id
                }, 500
    
    return decorated_function

def log_websocket_errors(f):
    """Decorator for WebSocket event error logging"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        try:
            return f(*args, **kwargs)
        except Exception as e:
            error_id = StructuredLogger.log_error(
                current_app.logger,
                e,
                context={
                    'websocket_event': f.__name__,
                    'args': args,
                    'kwargs': kwargs,
                    'socket_id': getattr(request, 'sid', None)
                }
            )
            
            # Emit error to client
            from flask_socketio import emit
            emit('error', {
                'code': 'INTERNAL_ERROR',
                'message': 'Internal server error',
                'error_id': error_id
            })
            
    return decorated_function

# Create logger instance
structured_logger = StructuredLogger()

File: mobile/src/utils/logger.js
/**
 * Mobile logging utilities with structured error reporting
 */

class MobileLogger {
  static generateId() {
    return Math.random().toString(36).substr(2, 9)
  }

  static logError(error, context = {}, component = 'Unknown') {
    const errorId = this.generateId()

    const errorData = {
      errorId,
      timestamp: new Date().toISOString(),
      component,
      platform: 'mobile',
      errorType: error.name || 'Error',
      errorMessage: error.message || String(error),
      stack: error.stack || 'No stack trace available',
      context,
      userId: context.userId || null
    }

    // Console error with detailed info
    console.group(`🔴 Mobile Error ${errorId} in ${component}`)
    console.error('Error:', error)
    console.log('Context:', context)
    console.log('Full Error Data:', errorData)
    console.groupEnd()

    // Send to backend in production (optional)
    if (__DEV__ === false && context.sendToBackend) {
      this.sendErrorToBackend(errorData).catch(console.warn)
    }

    return errorId
  }

  static logWarning(message, context = {}, component = 'Unknown') {
    const warningData = {
      id: this.generateId(),
      timestamp: new Date().toISOString(),
      component,
      platform: 'mobile',
      message,
      context
    }

    console.group(`🟡 Mobile Warning in ${component}`)
    console.warn(message)
    console.log('Context:', context)
    console.groupEnd()

    return warningData.id
  }

  static logInfo(message, context = {}, component = 'Unknown') {
    const infoData = {
      id: this.generateId(),
      timestamp: new Date().toISOString(),
      component,
      platform: 'mobile',
      message,
      context
    }

    console.group(`ℹ️ Mobile Info from ${component}`)
    console.log(message)
    console.log('Context:', context)
    console.groupEnd()

    return infoData.id
  }

  static logWebSocketEvent(event, data = {}, component = 'WebSocket') {
    const eventData = {
      id: this.generateId(),
      timestamp: new Date().toISOString(),
      component,
      platform: 'mobile',
      event,
      data,
      connectionState: data.socket?.connected ? 'connected' : 'disconnected'
    }

    console.group(`📡 Mobile WebSocket ${event} in ${component}`)
    console.log('Event Data:', data)
    console.log('Connection State:', eventData.connectionState)
    console.groupEnd()

    return eventData.id
  }

  static logNetworkRequest(url, method, statusCode, context = {}) {
    const requestData = {
      id: this.generateId(),
      timestamp: new Date().toISOString(),
      platform: 'mobile',
      url,
      method,
      statusCode,
      context
    }

    const statusEmoji = statusCode >= 400 ? '🔴' : statusCode >= 300 ? '🟡' : '🟢'

    console.group(`${statusEmoji} Mobile API ${method} ${url} (${statusCode})`)
    console.log('Request Data:', context)
    console.groupEnd()

    return requestData.id
  }

  static async sendErrorToBackend(errorData) {
    try {
      // This would need to be implemented based on your API client
      const response = await fetch('/api/client-errors', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(errorData)
      })
    } catch (backendError) {
      console.warn('Failed to send mobile error to backend:', backendError)
    }
  }
}

// Error boundary helper
export function withErrorLogging(fn, context = {}, component = 'Unknown') {
  return async (...args) => {
    try {
      return await fn(...args)
    } catch (error) {
      const errorId = MobileLogger.logError(error, {
        ...context,
        functionArgs: args
      }, component)

      // Re-throw with error ID
      error.errorId = errorId
      throw error
    }
  }
}

// API call wrapper
export function withApiErrorLogging(apiCall, context = {}, component = 'API') {
  return withErrorLogging(apiCall, context, component)
}

// WebSocket event wrapper
export function withWebSocketErrorLogging(handler, context = {}, component = 'WebSocket') {
  return (data) => {
    try {
      MobileLogger.logWebSocketEvent('event_received', { data, ...context }, component)
      return handler(data)
    } catch (error) {
      MobileLogger.logError(error, {
        ...context,
        eventData: data
      }, component)
      throw error
    }
  }
}

export default MobileLogger
